### DISCUSSION TITLE
Install packages before lock maintenance

### DISCUSSION THREAD

**Discussion by CedricCabessa** [renovatebot/renovate#23500] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

36.17.0

### If you're self-hosting Renovate, select which platform you are using.

GitHub Enterprise Server

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

Hi,

I'm using renovate to update a `Pipenv.lock` file.
My problem is that in order to `pipenv lock` to succeed I need to install some packages on the system (`build-essential` and `libpq-dev`).

Without those, `pipenv lock` fail and no PR is created.

How can I tell renovate to `apt install` those package *before* running `pipenv lock` ?

### What you tried so far.

I tried to open a shell on renovate docker image to play with things

* I cannot install libpq-dev via install-tools
* By default default, renovate doesn't run as root, so I cannot `apt install`

The only option I see is to build a custom docker image based on the official renovate but with extra package. I'd like to avoid this extra maintenance.

### Relevant debug logs

Pipenv lock failure (not related to renovate code, just a proof it fail :-) )

```
['/usr/bin/python3', '-c', '\nexec(compile(\'\'\'\n# This is <pip-setuptools-caller> -- a caller that pip uses to run setup.py\n#\n# - It imports setuptools
before invoking setup.py, to enable projects that directly\n#   import from `distutils.core` to work with newer packaging standards.\n# - It provides a 
clear error message when setuptools is not installed.\n# - It sets `sys.argv[0]` to the underlying `setup.py`, when invoking `setup.py` so\n#   setuptools 
doesn\'t think the script is `-c`. This avoids the following warning:\n#     manifest_maker: standard file \'-c\' not found".\n# - It generates a shim 
setup.py, for handling setup.cfg-only projects.\nimport os, sys, tokenize\n\ntry:\n    import setuptools\nexcept ImportError as error:\n    print(\n        
"ERROR: Can not execute `setup.py` since setuptools is not available in "\n        "the build environment.",\n        file=sys.stderr,\n    )\n    
sys.exit(1)\n\n__file__ = %r\nsys.argv[0] = __file__\n\nif os.path.exists(__file__):\n    filename = __file__\n    with tokenize.open(__file__) as f:\n     
setup_py_code = f.read()\nelse:\n    filename = "<auto-generated setuptools caller>"\n    setup_py_code = "from setuptools import setup; 
setup()"\n\nexec(compile(setup_py_code, filename, "exec"))\n\'\'\' % (\'/tmp/pip-temp-toj9a83g/psycopg2_67ad1fb419134476ba5ccbbc4b652c66/setup.py\',), 
"<pip-setuptools-caller>", "exec"))', '--no-user-cfg', 'egg_info', '--egg-base', '/tmp/pip-pip-egg-info-qb23t0uu']
/tmp/pip-temp-toj9a83g/psycopg2_67ad1fb419134476ba5ccbbc4b652c66/
⠸ Locking...ERROR:pip.subprocessor:[present-rich] python setup.py egg_info exited with 1
```

**Comment by rarkins** [comment]
If the installation requires root then:
- run as root, or
- build a custom image, or
- we build it into everyone's image

**Reply by CedricCabessa** [reply]
@rarkins thanks for the answer
Even if I run as root, how can run a command (apt or anything) before the "lock maintenance" is done ?

**Reply by rarkins** [reply]
If you are willing to run your containers as root, then it means the (future) feature preUpgradeTasks could solve this. There's a feature request issue open for it you can subscribe to if so

**Reply by CedricCabessa** [reply]
ok, yes I've seen this issue already on preUpgradeTasks ;-)

In the meanwhile, I have a small hack that solve my problem
I use a simple bash script `entrypoint.sh`

```
#!/bin/bash

apt update

apt install -y build-essential libpq-dev

runuser -u ubuntu renovate
```
and start renovate via docker

```
docker run  --env RENOVATE_TOKEN=$GH_TOKEN --env RENOVATE_CONFIG_FILE=/github-action/renovate-config.js --volume $PWD/.github/renovate-config.js:/github-action/renovate-config.js --volume /tmp:/tmp --volume $PWD/entrypoint.sh:/entrypoint.sh --rm  --user=0 ghcr.io/renovatebot/renovate:latest /entrypoint.sh
```

thx !
